### PR TITLE
MAINT: refactor _savez

### PR DESCRIPTION
**Pull request by kain88-de**
use context manager to open/close file handle.

### CODE REVIEW

**Comment by pv** [comment]
That fails to flush buffers.

**Comment by kain88-de** [comment]
I moved the call to `zip.writef`  out of the context manager. This should solve the flush error and ensures we only call `writef` if we could write the tmpfile like before.

**Comment by charris** [comment]
Needs rebase. Note that there are now two paths that need changing.

**Comment by charris** [comment]
See also #6545.

**Comment by charris** [comment]
Note the `temppath` context manager in `numpy/testing/`
```
In [1]: from numpy.testing import temppath

In [2]: temppath?
Type:        function
String form: <function temppath at 0x7f92bb9560c8>
File:        /usr/lib64/python2.7/contextlib.py
Definition:  temppath(*args, **kwds)
Docstring:
Context manager for temporary files.

Context manager that returns the path to a closed temporary file. Its
parameters are the same as for tempfile.mkstemp and are passed directly
to that function. The underlying file is removed when the context is
exited, so it should be closed at that time.

Windows does not allow a temporary file to be opened if it is already
open, so the underlying file must be closed after opening before it
can be opened again.
```

File: numpy/lib/npyio.py [hunk 1]
```diff
@@ -698,30 +698,21 @@ def _savez(file, args, kwds, compress, allow_pickle=True, pickle_kwargs=None):
...
-                        fid.close()
-        finally:
-            os.remove(tmpfile)
+                with open(tmpfile, 'wb') as fid:
+                    try:
+                        format.write_array(fid, np.asanyarray(val),
+                                           allow_pickle=allow_pickle,
+                                           pickle_kwargs=pickle_kwargs)
+                    except IOError as exc:
+                        raise IOError("Failed to write to %s: %s" % (tmpfile, exc))
```
Comment (kain88-de) [inline numpy/lib/npyio.py]: any policy on string interpolation while I'm cleaning up this part?

**Comment by kain88-de** [comment]
Is this still of interest?

**Pull request closed without merging**